string_indexing/suffix_array.py

```diff
@@ -63,6 +66,223 @@ def induced_sorting(text, n):
   '''Computes suffix array using Nong-Zhang-Chan algorithm'''
   raise NotImplementedError
 
+def _map_string(s):
+  sorted_char_set = sorted(set( c for c in s[1:] ))
+  char_to_ord = { c : i for i, c in enumerate(sorted_char_set, 1) }
+  return [0] + [ char_to_ord[c] for c in s[1:] ] + [0]
+
+def small_large(text, n):
+  Category = enum.Enum("Category", "L S")
+
+  def _small_large(s):
+    A, SL, N = _bucket_array(s), _categorize_sl(s), len(s)-1
+    if N == 2:
+      return [2, 1]
+    if SL.count(Category.L) == 1:
+      return [N] + list(range(1, N))
+    if SL.count(Category.S) == 1:
+      return list(reversed(range(1, N + 1)))
+
+    if SL[-1] == Category.S:
+      sortedS = _sort_category(s, SL, Category.S)
+      groups = A.groups.copy()
+      _move_to_group_end(
+          reversed(sortedS[1:]),
+          lambda x: True, False, A, groups)
+      _move_to_group_end(
+          map(lambda i : A.tab[i] - 1, range(1, len(A.tab))),
+          lambda x: SL[x] == Category.L, True, A, groups)
+    elif SL[-1] == Category.L:
+      sortedL = _sort_category(s, SL, Category.L)
+      groups = A.groups.copy()
+      _move_to_group_end(sortedL[1:], lambda x: True, True, A, groups)
+      _move_to_group_end(
+          map(lambda i : A.tab[i] - 1, reversed(range(1, len(A.tab)))),
+          lambda x: SL[x] == Category.S, False, A, groups)
+    else:
+      raise Exception("fail")
+
+    return A.tab[1:]
+
+  def _sort_category(s, SL, category):
+    suff_by_dist, categ_substr = _prepare_substr_lists(
+        _bucket_array(s), SL, category)
+
+    # sort categ_substr
+    for i in range(1, len(suff_by_dist)):
+      groups = suff_by_dist[i].sorted_groups()
+      for g in groups if category == Category.S else reversed(groups):
+        to_split = collections.defaultdict(list)
+        for j in suff_by_dist[i].tab[g[0]:g[1]+1]:
+          x = j-i
+          g_id = categ_substr.group_id[categ_substr.rev[x]]
+          to_split[g_id].append(x)
+
+        for g_id in to_split:
+          vals = to_split[g_id]
+          if category == Category.S:
+            categ_substr.split_group_before(vals)
+            categ_substr.reverse_group(
```

Comment:
If I'm not mistaken, this is common for both if and else, so we may put it after if/else clause.